File: 2019/Thang05_2019/16_05_2019/expression.cpp
#include <bits/stdc++.h>
#define forn(i, a, b) for (int i=(a); i<=(b); ++i)
#define ford(i, a, b) for (int i=(a); i>=(b); --i)
using namespace std;
//////

const
    int maxn=110, maxg=maxn*maxn;
int n, k;
int a[maxn];
bool f[maxn][2*maxg+1];
int d[maxn];
//////

void enter()
{
    scanf("%d%d\n", &n, &k);
    forn(i, 1, n) scanf("%d", &a[i]);
}
//////

void process()
{
    f[1][a[1]+maxg]=true;
    forn(i, 1, n-1)
        forn(j, 0, 2*maxg)
        if (f[i][j])
        {
            if (j+a[i+1]<=2*maxg) f[i+1][j+a[i+1]]=true;
            if (j-a[i+1]>=0) f[i+1][j-a[i+1]]=true;
        }
    int i=n, j=k+maxg;
    while (i>1)
    {
        if (j-a[i]>=0 && f[i-1][j-a[i]])
        {
            d[i]=1;





            j-=a[i];
        } else
        {
            d[i]=-1;
            j+=a[i];
        }
        i--;
    }
    forn(i, 2, n)
        if (d[i]==1) cout << '+'; else cout << '-';
}
//////

int main()
{
    freopen("expression.inp", "r", stdin);
    freopen("expression.out", "w", stdout);
    enter();
    process();
}

File: 2018/Thang11-2018/08-11-2018/expression.pas
Const   ginp='expression.inp';
        gout='expression.out';
        maxn=101;
        maxg=10001;

Var     n,gg:longint;
        a:array[0..maxn] of longint;
        d:array[0..maxn,-maxg..maxg] of boolean;
        res:string;

Procedure Enter;
Var     i:longint;
Begin
        readln(n,gg);
        for i:=1 to n do read(a[i]);
End;

Procedure Process;
Var     i,j:longint;
Begin
        fillchar(d,sizeof(d),false);
        d[1,a[1]]:=true;
        for i:=2 to n do
                for j:=-maxg to maxg do
                        if ((j-a[i]>=-maxg) and (d[i-1,j-a[i]])) or ((j+a[i]<=maxg) and (d[i-1,j+a[i]])) then d[i,j]:=true;
        res:='';
        j:=gg;
        for i:=n downto 2 do
                if (j-a[i]>=-maxg) and (d[i-1,j-a[i]]) then
                begin
                        res:='+'+res;
                        j:=j-a[i];
                end else
                begin
                        res:='-'+res;
                        j:=j+a[i];
                end;
        write(res);
End;

Begin
        Assign(input,ginp); Assign(output,gout);
        Reset(input); Rewrite(output);
        Enter;
        Process;
        Close(input); Close(output);
End.
